browser/ui/tabs/brave_tab_menu_model.cc

```diff
@@ -0,0 +1,64 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/tabs/brave_tab_menu_model.h"
+
+#include "brave/browser/ui/tabs/tab_menu_model_delegate_proxy.h"
+#include "chrome/browser/sessions/tab_restore_service_factory.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "chrome/browser/ui/tabs/tab_strip_model_delegate.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/sessions/core/tab_restore_service.h"
+
+BraveTabMenuModel::BraveTabMenuModel(ui::SimpleMenuModel::Delegate* delegate,
+                                     TabStripModel* tab_strip_model,
+                                     int index)
+    : TabMenuModel(new TabMenuModelDelegateProxy(delegate,
+                                                 tab_strip_model,
+                                                 index),
+                   tab_strip_model,
+                   index),
+      index_(index),
+      tab_strip_model_(tab_strip_model) {
+  // Take owner ship of TabMenuModelDelegateProxy instance.
+  // There is no way to instantiate it before calling ctor of base class.
+  delegate_proxy_.reset(TabMenuModel::delegate());
```

Comment:
is this needed anymore? (how is `delegate_proxy_` used?)

Reply:
It needs because `SimpleMenuModel` doesn't own delegate.
`BraveTabMenuModel` should manage proxy's lifecycle.

Reply:
This definitely does not look right to me. The TabMenuModel is not supposed to take ownership of the delegate and this seems likely to cause a crash at some point

Reply:
The delegate is owned by the BrowserTabStripController and I'm fairly certain that this will crash if you close a window with a context menu open. At the very least it breaks the cancel behavior here
```
BrowserTabStripController::~BrowserTabStripController() {
  // When we get here the TabStrip is being deleted. We need to explicitly
  // cancel the menu, otherwise it may try to invoke something on the tabstrip
  // from its destructor.
  if (context_menu_contents_.get())
    context_menu_contents_->Cancel();

  model_->RemoveObserver(this);
}
```

Reply:
so I just noticed that you're creating a new proxy delegate there, but that's also a potential problem. Let's dm and discuss when you're available

Reply:
I feel like this creates a bit of a convoluted chain of ownership and while it may or may not be ok, it seems like it would be easier to subclass BrowserTabStripController in BrowserViews and implement `ShowContextMenuForTab`. Then we can create our own delegate from the start with our own `TabMenuModel` and calling back into our own `ExecuteCommandForTab`. Then we can check the command ids in our BrowserTabStripController and either delegate to the TabStripModel or call the command in `BraveBrowserTabStripController`

Reply:
then at least the commands have the same flow at least through `ExecuteCommandForTab`
and there are no confusing ownership chains